File: dft/post-vv10/vv10.py
#!/usr/bin/env python

import os, time, numpy, ctypes
from pyscf import gto, scf, lib, dft

_loaderpath = os.path.dirname(__file__)
libvv10 = numpy.ctypeslib.load_library('libvv10.so', _loaderpath)

mol = gto.Mole()
mol.verbose = 4
mol.atom = '''
O
H 1 1.1
H 1 1.1 2 104.0
'''
mol.basis = 'aug-cc-pvdz'
mol.symmetry = 1
mol.build()

mf = dft.RKS(mol)
mf.xc = 'rpw86,pbe'
mf.grids.level = 1
mf.kernel()

dm = mf.make_rdm1()
coords = mf.grids.coords
weights = mf.grids.weights
ngrids = len(weights)
ao = dft.numint.eval_ao(mol, coords, deriv=1)
rho = dft.numint.eval_rho(mol, ao, dm, xctype='GGA')
gnorm2 = numpy.zeros(ngrids)
for i in range(ngrids):
    gnorm2[i] = numpy.linalg.norm(rho[-3:,i])**2
lib.logger.info(mf,'Rho = %.12f' % numpy.einsum('i,i->', rho[0], weights))
ex, vx = dft.libxc.eval_xc('rPW86,', rho)[:2]
ec, vc = dft.libxc.eval_xc(',PBE', rho)[:2]
lib.logger.info(mf, 'Exc = %.12f' % numpy.einsum('i,i,i->', ex+ec, rho[0], weights))

coef_C = 0.0093
coef_B = 5.9
coef_beta = 1.0/32.0 * (3.0/(coef_B**2.0))**(3.0/4.0)
kappa_pref = coef_B * (1.5*numpy.pi)/((9.0*numpy.pi)**(1.0/6.0))
const = 4.0/3.0 * numpy.pi
vv10_e = 0.0

t = time.time()
for idx1 in range(ngrids):
    point1 = coords[idx1,:]
    rho1 = rho[0,idx1]
    weigth1 = weights[idx1]
    gamma1 = gnorm2[idx1]
    Wp1 = const*rho1
    Wg1 = coef_C * ((gamma1/(rho1*rho1))**2.0)
    W01 = numpy.sqrt(Wg1 + Wp1)
    kappa1 = rho1**(1.0/6.0)*kappa_pref
    #
    R =  (point1[0]-coords[:,0])**2
    R += (point1[1]-coords[:,1])**2
    R += (point1[2]-coords[:,2])**2
    Wp2 = const*rho[0]
    Wg2 = coef_C * ((gnorm2/(rho[0]*rho[0]))**2.0)
    W02 = numpy.sqrt(Wg2 + Wp2)
    kappa2 = rho[0]**(1.0/6.0)*kappa_pref
    g = W01*R + kappa1
    gp = W02*R + kappa2
    kernel12 = -1.5*weights*rho[0]/(g*gp*(g+gp))

    # Energy 
    kernel = kernel12.sum()
    vv10_e += weigth1*rho1*(coef_beta + 0.5*kernel)
lib.logger.info(mf, 'VV10 = %.12f' % vv10_e)
lib.logger.info(mf, 'Total time taken VV10: %.3f seconds' % (time.time()-t))
   
t = time.time()
libvv10.vv10.restype = ctypes.c_double
ev = libvv10.vv10(ctypes.c_int(ngrids),
             ctypes.c_double(coef_C),
             ctypes.c_double(coef_B),
             coords.ctypes.data_as(ctypes.c_void_p),
             rho.ctypes.data_as(ctypes.c_void_p),
             weights.ctypes.data_as(ctypes.c_void_p),
             gnorm2.ctypes.data_as(ctypes.c_void_p))
lib.logger.info(mf, 'VV10 = %.12f' % ev)
lib.logger.info(mf, 'Total time taken VV10 (C): %.3f seconds' % (time.time()-t))

File: dft/post-vv10/vv10.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double vv10(const int n, const double coef_C,
                         const double coef_B,
                         const double *coords, 
                         const double *rho,
                         const double *weights,
                         const double *gnorm2)
{

  const double pi = 3.14159265358979323846264338328; 
  //const double coef_C = 0.0093;
  //const double coef_B = 5.9;
  const double const43 = 4.0/3.0*pi;

  size_t idx, jdx;
  double vv10_e = 0.0;

  double coef_beta;
  coef_beta = 1.0/32.0*pow(3.0/(coef_B*coef_B),3.0/4.0);
  double kappa_pref;
  kappa_pref = coef_B*(1.5*pi)/pow(9.0*pi,1.0/6.0);

  for (idx=0; idx<n; idx++){
    double point1x = coords[idx*3+0];
    double point1y = coords[idx*3+1];
    double point1z = coords[idx*3+2];
    double rho1 = rho[idx];
    double weigth1 = weights[idx];
    double gamma1 = gnorm2[idx];
    double Wp1 = const43*rho1;
    double Wg1 = coef_C*pow(gamma1/(rho1*rho1),2);
    double W01 = sqrt(Wg1 + Wp1);
    double kappa1 = pow(rho1,1.0/6.0)*kappa_pref;
    double kernel = 0.0;
#pragma omp parallel default(none) reduction(+:kernel) \
    shared(point1x,point1y,point1z,coords,rho,weights,kappa_pref,gnorm2,kappa1,W01)
{
#pragma omp for nowait schedule(dynamic)
    for (jdx=0; jdx<n; jdx++){
      double R = (point1x-coords[jdx*3+0])*(point1x-coords[jdx*3+0]);
      R += (point1y-coords[jdx*3+1])*(point1y-coords[jdx*3+1]);
      R += (point1z-coords[jdx*3+2])*(point1z-coords[jdx*3+2]);
      double rho2 = rho[jdx];
      double weigth2 = weights[jdx];
      double gamma2 = gnorm2[jdx];
      double Wp2 = const43*rho2;
      double Wg2 = coef_C*pow(gamma2/(rho2*rho2),2);
      double W02 = sqrt(Wg2 + Wp2);
      double kappa2 = pow(rho2,1.0/6.0)*kappa_pref;
      double g = W01*R + kappa1;
      double gp = W02*R + kappa2;
      kernel += -1.5*weigth2*rho2/(g*gp*(g+gp));
    }
} 
    vv10_e += weigth1*rho1*(coef_beta + 0.5*kernel);
  }
 
  return vv10_e;

}
